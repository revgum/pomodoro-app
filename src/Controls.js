// @flow
import * as React from "react";
import "./Controls.css";

// material-ui
import { withStyles } from "material-ui/styles";
import IconButton from "material-ui/IconButton";
import PlayCircleOutline from "@material-ui/icons/PlayCircleOutline";
import AlarmOff from "@material-ui/icons/AlarmOff";
import PauseCircleOutline from "@material-ui/icons/PauseCircleOutline";
import RepeatOne from "@material-ui/icons/RepeatOne";

type Props = {
  classes: any,
  clickPause: Function,
  clickReset: Function,
  clickStartStop: Function,
  isPaused: boolean,
  isStarted: boolean,
  secondsRemaining: number
};

const styles = theme => ({
  button: {
    margin: theme.spacing.unit
  },
  input: {
    display: "none"
  }
});

class Controls extends React.Component<Props> {
  render() {
    return (
      <div className="timer-controls">
        {this.startStopButton()}
        <IconButton
          aria-label="Pause"
          className={this.props.classes.button}
          disabled={this.props.isPaused || !this.props.isStarted}
          onClick={this.props.clickPause}
        >
          <PauseCircleOutline />
        </IconButton>
        <IconButton
          aria-label="Restart"
          className={this.props.classes.button}
          disabled={!this.props.isStarted}
          onClick={this.props.clickReset}
        >
          <RepeatOne />
        </IconButton>
      </div>
    );
  }

  startStopButton = () => {
    if (this.props.isStarted) {
      return (
        <IconButton
          aria-label="Stop"
          className={this.props.classes.button}
          onClick={this.props.clickStartStop}
        >
          <AlarmOff />
        </IconButton>
      );
    } else {
      return (
        <IconButton
          aria-label="Start"
          className={this.props.classes.button}
          onClick={this.props.clickStartStop}
        >
          <PlayCircleOutline />
        </IconButton>
      );
    }
  };
}

export default withStyles(styles)(Controls);
