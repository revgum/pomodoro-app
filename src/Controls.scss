@import "styles/shared.scss";

.timer-controls {
  ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}
