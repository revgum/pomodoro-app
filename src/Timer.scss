@import "styles/shared.scss";

.timer {
  .remaining {
    font-size: 4rem;
    text-align: center;
  }
}
