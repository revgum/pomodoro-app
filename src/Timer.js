// @flow
import * as React from "react";
import "./Timer.css";
import Controls from "./Controls";
import PauseForm from "./PauseForm";
import PauseList from "./PauseList";
import type { PauseState, TimerState } from "./types";
import * as Utils from "./utils";

type Props = {
  completeSession: Function,
  minutes: number,
  seconds: number
};

export default class Timer extends React.Component<Props, TimerState> {
  static defaultProps = {
    minutes: 25,
    seconds: 0
  };

  state: TimerState = {
    isPaused: false,
    isStarted: true,
    minutes: 25,
    pauses: [],
    pausedElapsed: 0,
    seconds: 0,
    secondsElapsed: 0,
    secondsRemaining: 25 * 60,
    timer: null
  };

  constructor(props: Props) {
    super(props);
    this.state = Object.assign(this.state, {
      minutes: props.minutes,
      seconds: props.seconds,
      secondsRemaining: props.minutes * 60 + props.seconds
    });
  }

  componentDidMount() {
    let timer_id = setInterval(this.tick, 1000);
    this.setState({
      timer: timer_id
    });
  }

  render() {
    return (
      <div className="timer">
        <div className="remaining">
          {Utils.formattedTime(this.state.secondsRemaining)}
        </div>
        <Controls
          clickPause={this.clickPause}
          clickReset={this.clickReset}
          clickStartStop={this.clickStartStop}
          isPaused={this.state.isPaused}
          isStarted={this.state.isStarted}
          secondsRemaining={this.state.secondsRemaining}
        />
        <PauseForm
          isPaused={this.state.isPaused}
          pausedElapsed={this.state.pausedElapsed}
          savePause={this.savePause}
        />
        <PauseList
          isPaused={this.state.isPaused}
          pauses={this.state.pauses}
          pausedElapsed={this.state.pausedElapsed}
        />
      </div>
    );
  }

  clickPause = () => {
    this.setState({
      isPaused: !this.state.isPaused
    });
  };

  clickReset = () => {
    if (this.state.timer !== null) clearInterval(this.state.timer);
    this.setState({
      isPaused: false,
      isStarted: false,
      timer: null,
      pauses: [],
      pausedElapsed: 0,
      secondsElapsed: 0,
      secondsRemaining: this.props.minutes * 60 + this.props.seconds
    });
  };

  /*
   * Stop (Finish) the pomodoro if it is currently running, otherwise
   * start the fresh pomodoro.
   */
  clickStartStop = () => {
    if (this.state.isStarted) {
      this.props.completeSession(this.state);
      this.clickReset();
    } else {
      let timer_id = setInterval(this.tick, 1000);
      this.setState({
        isPaused: false,
        isStarted: true,
        timer: timer_id
      });
    }
  };

  savePause = (description: string) => {
    let paused: PauseState = {
      seconds: this.state.pausedElapsed,
      description: description
    };

    this.state.pauses &&
      this.setState({
        isPaused: false,
        pausedElapsed: 0,
        pauses: [paused, ...this.state.pauses]
      });
  };

  /*
   * With each tick of the timer, adjust the appropriate counters or if the
   * time has ran out then finish the pomodoro.
   */
  tick = () => {
    let remaining = this.state.secondsRemaining;
    if (remaining <= 0) {
      this.clickStartStop();
    } else if (this.state.isStarted && !this.state.isPaused) {
      this.setState(prevState => ({
        secondsElapsed: prevState.secondsElapsed + 1,
        secondsRemaining: prevState.secondsRemaining - 1
      }));
    } else if (this.state.isPaused) {
      this.setState(prevState => ({
        pausedElapsed: prevState.pausedElapsed + 1,
        secondsElapsed: prevState.secondsElapsed + 1
      }));
    }
  };
}
