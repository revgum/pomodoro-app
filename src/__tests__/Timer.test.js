import React from "react";
import Timer from "../Timer";
import Controls from "../Controls";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

// Shallow rendering test
it("passes a smoke test", () => {
  shallow(<Timer />);
});

it("captures a snapshot", () => {
  const tree = renderer.create(<Timer />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("sets a timer interval for counting down the timer", () => {
  jest.useFakeTimers();
  const wrapper = shallow(<Timer />);
  expect(setInterval).toHaveBeenCalledTimes(1);
  expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
});

it("defaults to a 25 minute timer", () => {
  const wrapper = shallow(<Timer />);
  expect(wrapper).toContainReact(<div className="remaining">25:00</div>);
});

it("renders a specified timer", () => {
  const wrapper = shallow(<Timer minutes={10} seconds={0} />);
  expect(wrapper).toContainReact(<div className="remaining">10:00</div>);

  // advance the timer by a tick
  wrapper.setState({ secondsRemaining: 599 });
  expect(wrapper).toContainReact(<div className="remaining">09:59</div>);
});

it("renders an hour timer", () => {
  const wrapper = shallow(<Timer minutes={60} seconds={0} />);
  expect(wrapper).toContainReact(<div className="remaining">1:00:00</div>);
});

describe("#tick", () => {
  it("decrements the timer and increments elapsed time", () => {
    const wrapper = shallow(<Timer minutes={10} seconds={0} />);
    const instance = wrapper.instance();
    const tickSpy = jest.spyOn(instance, "tick");
    instance.tick();
    expect(tickSpy).toHaveBeenCalled();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: true,
      secondsElapsed: 1,
      secondsRemaining: 599
    });
  });
  it("stops and resets the timer when it reaches 0", () => {
    const wrapper = shallow(<Timer minutes={0} seconds={1} />);
    const instance = wrapper.instance();
    const tickSpy = jest.spyOn(instance, "tick");

    // Tick the timer for the final second
    instance.tick();
    expect(tickSpy).toHaveBeenCalled();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: true,
      secondsElapsed: 1,
      secondsRemaining: 0
    });

    // Tick the timer after time has expired
    instance.tick();
    expect(tickSpy).toHaveBeenCalled();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: false,
      secondsElapsed: 0,
      secondsRemaining: 0,
      timer: null
    });
  });
});

describe("#clickPause", () => {
  const wrapper = shallow(<Timer minutes={0} seconds={1} />);
  const instance = wrapper.instance();
  const tickSpy = jest.spyOn(instance, "tick");
  const pauseSpy = jest.spyOn(instance, "clickPause");

  it("pauses the timer and increments the elapsed time", () => {
    expect(wrapper).toHaveState({
      isPaused: false
    });

    instance.clickPause();
    expect(pauseSpy).toHaveBeenCalled();
    instance.tick();
    expect(wrapper).toHaveState({
      isPaused: true,
      secondsElapsed: 1
    });
  });
});

describe("#clickReset", () => {
  const wrapper = shallow(<Timer minutes={0} seconds={1} />);
  const instance = wrapper.instance();
  const tickSpy = jest.spyOn(instance, "tick");
  const resetSpy = jest.spyOn(instance, "clickReset");

  it("resets the timer", () => {
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: true,
      secondsElapsed: 0,
      secondsRemaining: 1
    });

    instance.clickReset();
    expect(resetSpy).toHaveBeenCalled();
    instance.tick();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: false,
      timer: null,
      secondsElapsed: 0,
      secondsRemaining: 1
    });
  });
});

describe("#clickStartStop", () => {
  const wrapper = shallow(<Timer minutes={0} seconds={1} />);
  const instance = wrapper.instance();
  const tickSpy = jest.spyOn(instance, "tick");
  const startStopSpy = jest.spyOn(instance, "clickStartStop");
  const resetSpy = jest.spyOn(instance, "clickReset");

  it("stops and resets a timer that is running", () => {
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: true,
      secondsElapsed: 0,
      secondsRemaining: 1
    });

    instance.clickStartStop();
    expect(startStopSpy).toHaveBeenCalled();
    expect(resetSpy).toHaveBeenCalled();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: false,
      timer: null,
      secondsElapsed: 0,
      secondsRemaining: 1
    });
  });

  it("starts a timer that was reset", () => {
    instance.clickReset();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: false,
      timer: null,
      secondsElapsed: 0,
      secondsRemaining: 1
    });

    instance.clickStartStop();
    expect(startStopSpy).toHaveBeenCalled();
    expect(wrapper).toHaveState({
      isPaused: false,
      isStarted: true,
      secondsElapsed: 0,
      secondsRemaining: 1
    });
  });
});
