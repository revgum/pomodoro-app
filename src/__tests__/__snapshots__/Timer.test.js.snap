// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captures a snapshot 1`] = `
<div
  className="timer"
>
  <div
    className="remaining"
  >
    25:00
  </div>
  <div>
    <ul>
      <li
        className="start started"
        onClick={[Function]}
      >
        Stop
      </li>
      <li
        className="pause enabled"
        onClick={[Function]}
      >
        Pause
      </li>
      <li
        className="reset"
        onClick={[Function]}
      >
        Reset
      </li>
    </ul>
  </div>
</div>
`;
